src/CentrifugeRouter.sol

```diff
@@ -64,30 +60,90 @@ contract CentrifugeRouter is Auth, Multicall, ICentrifugeRouter {
+        uint256 lockedRequest = lockedRequests[controller][vault];
+        require(lockedRequest > 0, "CentrifugeRouter/controller-has-no-balance");
+        lockedRequests[controller][vault] = 0;
+        IERC7540(vault).requestDeposit(lockedRequest, controller, address(this));
+        emit ExecuteLockedDepositRequest(vault, controller);
     }
 
     /// @inheritdoc ICentrifugeRouter
-    function claimDeposit(address vault, address user) external {
-        uint256 maxDeposit = IERC7540(vault).maxDeposit(user);
-        IERC7540(vault).deposit(maxDeposit, user, user);
+    function claimDeposit(address vault, address receiver, address controller) external {
+        uint256 maxDeposit = IERC7540(vault).maxDeposit(controller);
+        IERC7540(vault).deposit(maxDeposit, receiver, controller);
     }
 
     // --- Redeem ---
     /// @inheritdoc ICentrifugeRouter
-    function requestRedeem(address vault, uint256 amount) external {
-        IERC7540(vault).requestRedeem(amount, msg.sender, msg.sender);
+    function requestRedeem(address vault, uint256 amount, address controller, address owner) external {
+        IERC7540(vault).requestRedeem(amount, controller, owner);
+    }
+
+    /// @inheritdoc ICentrifugeRouter
+    function claimRedeem(address vault, address receiver, address controller) external {
+        uint256 maxRedeem = IERC7540(vault).maxRedeem(controller);
+        IERC7540(vault).redeem(maxRedeem, receiver, controller);
+    }
+
+    // --- ERC20 approval ---
+    /// @inheritdoc ICentrifugeRouter
+    function approveMax(address token, address spender) external {
+        if (IERC20(token).allowance(address(this), spender) == 0) {
+            SafeTransferLib.safeApprove(token, spender, type(uint256).max);
+        }
     }
 
     /// @inheritdoc ICentrifugeRouter
-    function claimRedeem(address vault, address user) external {
-        uint256 maxRedeem = IERC7540(vault).maxRedeem(user);
-        IERC7540(vault).redeem(maxRedeem, user, user);
+    function permit(address asset, uint256 assets, uint256 deadline, uint8 v, bytes32 r, bytes32 s) external {
+        try IERC20Permit(asset).permit(msg.sender, address(this), assets, deadline, v, r, s) {} catch {}
+    }
+
+    // --- ERC20 wrapping ---
+    /// @dev requires calling approveMax(underlying, wrapper) before
+    function wrap(address wrapper, uint256 amount, address receiver) external {
+        address underlying = IERC20Wrapper(wrapper).underlying();
+
+        amount = MathLib.min(amount, IERC20(underlying).balanceOf(msg.sender));
+        require(amount != 0, "CentrifugeRouter/zero-balance");
+        SafeTransferLib.safeTransferFrom(underlying, msg.sender, address(this), amount);
+
+        require(IERC20Wrapper(wrapper).depositFor(receiver, amount), "CentrifugeRouter/deposit-for-failed");
+    }
+
+    function unwrap(address wrapper, uint256 amount, address receiver) external {
+        require(receiver != address(0), "CentrifugeRouter/zero-address");
```

Comment:
Does this add the necessary security ? There was this notion that we can still pass wrong address

Reply:
I think we can possibly even remove this require, it doesn't seem to add any security. Really just adapted from Morpho but I can't think of a good reason (besides an auditor commenting that there should be address zero checks..)